(CDAP-1741) Docs for app logbacks

ParthGandhi opened the pull request:
Re-opening https://github.com/caskdata/cdap/pull/1597 against release with addressed comments.

poornachandra commented on cdap-docs/developers-manual/source/advanced/application-logback.rst:
```diff
@@ -0,0 +1,43 @@
+Application Logbacks
+====================
+
+YARN containers spawned by a CDAP application use the default container logback, ``logback-container.xml``,
```
YARN containers (both master services and user programs) launched by CDAP by default use the logback configuration file present in :ref:`Configuration <install-configuration>` directory —  logback-container.xml.

poornachandra replied:
Should we write another section for master services logback? It does not seem to fit in this section. Maybe in operations section?

awholegunch replied:
We could, but in lieu, let's add a glossary entry to ``cdap-docs/reference-manual/source/glossary.rst``.
Please edit/correct this as appropriate:

    Master Services
         The YARN container with the system services started by CDAP.

awholegunch replied:
And then in the first comment, we can reference it with

    :term:`master services`

and that should link to the Glossary.

awholegunch replied:
For the first line then:

    YARN containers (both :term:`master services` and user programs) launched by CDAP by default use the 
    logback configuration file present in the :ref:`Configuration <install-configuration>` directory:  
    ``logback-container.xml``.

Except it looks like standalone uses "logbook.xml".

poornachandra commented on cdap-docs/developers-manual/source/advanced/application-logback.rst:
```diff
@@ -0,0 +1,43 @@
+====================
+
+YARN containers spawned by a CDAP application use the default container logback, ``logback-container.xml``,
+packaged with CDAP. This ``logback.xml`` does log rotation, and depending on the use case, may be sufficient.
```
``logback-container.xml`` has configuration to rotate logs once every day and delete logs older than 14 days.

poornachandra commented on cdap-docs/developers-manual/source/advanced/application-logback.rst:
```diff
@@ -0,0 +1,43 @@
+YARN containers spawned by a CDAP application use the default container logback, ``logback-container.xml``,
+packaged with CDAP. This ``logback.xml`` does log rotation, and depending on the use case, may be sufficient.
+However, it is also possible to specify a custom ``logback.xml`` for a CDAP application.
+The packaged ``logback.xml`` is then used for each container spawned by a program of the application.
```
The packaged ``logback.xml`` is then used for each container launched for the application.

poornachandra commented on cdap-docs/developers-manual/source/advanced/application-logback.rst:
```diff
@@ -0,0 +1,43 @@
+
+YARN containers spawned by a CDAP application use the default container logback, ``logback-container.xml``,
+packaged with CDAP. This ``logback.xml`` does log rotation, and depending on the use case, may be sufficient.
+However, it is also possible to specify a custom ``logback.xml`` for a CDAP application.
```
However, it is also possible to use a custom ``logback.xml`` for a CDAP application by packaging it with the application in ``src/main/resources`` directory.

awholegunch replied:
How about:
```
for a CDAP application by packaging it with the application in the application's ``src/main/resources`` directory.
```

poornachandra commented on cdap-docs/developers-manual/source/advanced/application-logback.rst:
```diff
@@ -0,0 +1,43 @@
+in the application jar file. The ``pom.xml`` of the application, with the application's ``resource`` directory
+containing the ``logback.xml``, might look like this::
+
+  <?xml version="1.0" encoding="UTF-8"?>
```
This is not required, right?

ParthGandhi replied:
Leaving as an example. Don't think it hurts.

poornachandra commented on cdap-docs/developers-manual/source/advanced/application-logback.rst:
```diff
@@ -0,0 +1,43 @@
+
+  </project>
+
+CDAP supports standard logback functionality. To write a custom logback, you may
```
CDAP supports standard logback functionality. -- I don't think this sentence is required.

poornachandra commented on cdap-docs/developers-manual/source/advanced/application-logback.rst:
```diff
@@ -0,0 +1,43 @@
+  </project>
+
+CDAP supports standard logback functionality. To write a custom logback, you may
+refer to `LOGBack <http://logback.qos.ch/>`__. It's important to note that when a ``logback.xml`` is
```
It would be a good idea to put a separate note section to highlight the note.

poornachandra replied:
It's important to note that when a custom ``logback.xml`` is specified for an application, the custom ``logback.xml`` will be used in place of ``logback-container.xml``. The custom ``logback.xml`` needs to have configuration for log rotation and log clean-up, to ensure that long-running containers don't fill up the disk.

poornachandra replied:
Also, it is Logback not LOGBack

awholegunch replied:
For notes, I usually start a new paragraph, like this (note the edits inside):
```
**Note:** When a custom ``logback.xml`` is specified for an application, the custom ``logback.xml`` will be used in place of ``logback-container.xml``. The custom ``logback.xml`` needs to be configured for log rotation and log clean-up to ensure that long-running containers don't fill up the disk.
```

ParthGandhi replied:
Okay, i'll make it a separate section.

I used the spelling from their logo: http://logback.qos.ch ...

poornachandra commented:
Done with one round of review. We need to add a section in operations for log handling of master services.

awholegunch commented on cdap-docs/developers-manual/source/advanced/application-logback.rst:
```diff
@@ -0,0 +1,43 @@
+packaged with CDAP. This ``logback.xml`` does log rotation, and depending on the use case, may be sufficient.
+However, it is also possible to specify a custom ``logback.xml`` for a CDAP application.
+The packaged ``logback.xml`` is then used for each container spawned by a program of the application.
+
```
insert a highlight directive:
```
.. highlight:: xml
```

poornachandra commented:
It would be a good idea to add a note in the upgrade section to copy `logback-container.xml` to conf dir.

awholegunch commented on cdap-docs/reference-manual/source/glossary.rst:
```diff
@@ -130,5 +130,9 @@ Glossary
       isolation.
+
+   Master Services
+      YARN containers started by the ``cdap-master`` process. This includes the containers that do logging,
```
that perform logging,

awholegunch commented on cdap-docs/admin-manual/source/operations/logback.rst:
```diff
@@ -0,0 +1,12 @@
+.. meta::
+:author: Cask Data, Inc.
```
add spaces: "    " to match line following.

ParthGandhi commented:
@poornachandra That should go only in 2.6.2 docs. And should also go in the upgrade docs written by Rohit & co for namespaces. It does not fit well into our current upgrade docs. 

@awholegunch @poornachandra Addressed comments. Please review.

awholegunch commented on cdap-docs/admin-manual/source/operations/logback.rst:
```diff
@@ -0,0 +1,12 @@
+
+:term:`Master Services <master services>` use ``logback-container.xml``, packaged with the CDAP distribution,
+for logging configuration. This ``logback`` does log rotation once a day at midnight and expires logs older than
+14 days. Changes can be made to ``logback-container.xml`` but the ``cdap-master`` process will need to be restarted
```
```
``logback-container.xml``, but the
```
add comma

awholegunch replied:
Do you think there is more that can be added to this? Otherwise, it seems short. Does it justify its own page? Could it be added to an existing instead?

poornachandra replied:
14 days. Changes can be made to ``logback-container.xml``, and the ``cdap-master`` process...

awholegunch commented:
A couple of comments. Need to resolve if this extra page should be added.

poornachandra commented on cdap-docs/developers-manual/source/advanced/application-logback.rst:
```diff
@@ -0,0 +1,48 @@
+
+  </project>
+
+To write a custom logback, you may refer to `LOGBack <http://logback.qos.ch/>`__.
```
Even though the logo might say LOGBack, the text says Logback. Since we are using the text here, let's use Logback here.

poornachandra commented on cdap-docs/developers-manual/source/advanced/application-logback.rst:
```diff
@@ -0,0 +1,29 @@
+it with the application in the application's ``src/main/resources`` directory.
+The packaged ``logback.xml`` is then used for each container launched by the application.
+
+Writing and packaging a logback.xml
```
Do we still need this section, since we already say that we can include `logback.xml` in `src/main/resources` directory above?

poornachandra commented on cdap-docs/reference-manual/source/glossary.rst:
```diff
@@ -130,5 +130,9 @@ Glossary
       isolation.
+
+   Master Services
+      YARN containers started by the ``cdap-master`` process. This includes the containers that perform logging,
```
CDAP system services that are run in YARN containers like Transaction Service, Dataset Executor, Log Saver, Metrics Processor, etc.

poornachandra commented:
Minor comments, otherwise LGTM :+1:

poornachandra commented:
:+1:
